packages/expo-screen-orientation/ios/EXScreenOrientation/EXScreenOrientationModule.m

```diff
@@ -0,0 +1,225 @@
+                    rejecter:(UMPromiseRejectBlock)reject)
+{
+  resolve([EXScreenOrientationUtilities exportOrientation:_currentScreenOrientation]);
+}
+
+// Will be called when this module's first listener is added.
+- (void)startObserving
+{
+  [[NSNotificationCenter defaultCenter] addObserver:self
+                                           selector:@selector(handleDeviceOrientationChange:)
+                                               name:UIDeviceOrientationDidChangeNotification
+                                             object:[UIDevice currentDevice]];
+  _hasListeners = YES;
+}
+
+// Will be called when this module's last listener is removed, or on dealloc.
+- (void)stopObserving
+{
+  [[NSNotificationCenter defaultCenter] removeObserver:self];
+  _hasListeners = NO;
+}
+
+- (void)handleDeviceOrientationChange:(NSNotification *)notification
+{
+  UIDevice *device = notification.object;
+  UIInterfaceOrientation currentDeviceOrientation = [EXScreenOrientationUtilities UIDeviceOrientationToUIInterfaceOrientation:[device orientation]];
+  UIInterfaceOrientationMask currentOrientationMask = [_screenOrienationRegistry currentOrientationMask];
+  
+  // checks if screen orientation should be changed when user rotates the device
+  if ([EXScreenOrientationUtilities doesOrientationMask:currentOrientationMask containOrientation:currentDeviceOrientation] && _currentScreenOrientation != currentDeviceOrientation) {
+    // updates the current screen orientation
+    _currentScreenOrientation = currentDeviceOrientation;
+    
+    // we sends event only if current screen orientation was changed
+    [self handleScreenOrientationChange];
+  }
+}
+
+- (void)handleScreenOrientationChange
+{
+  if (_hasListeners) {
+    [_eventEmitter sendEventWithName:@"expoDidUpdateDimensions" body:@{
+      @"orientation": [EXScreenOrientationUtilities exportOrientation:_currentScreenOrientation],
+      @"orientationLock": [EXScreenOrientationUtilities exportOrientationLock:[_screenOrienationRegistry currentOrientationMask]]
+    }];
+  }
+}
+
+- (NSArray<NSString *> *)supportedEvents {
+  return @[@"expoDidUpdateDimensions"];
+}
+
+- (void)enforceDesiredDeviceOrientationWithOrientationMask:(UIInterfaceOrientationMask)orientationMask
+{
+  // if current sreen orientation isn't part of the mask, we have to change orientation for default one included in mask, in order up-left-right-down
+  if (![EXScreenOrientationUtilities doesOrientationMask:orientationMask containOrientation:_currentScreenOrientation]) {
+    __block UIInterfaceOrientation newOrientation = [EXScreenOrientationUtilities defaultOrientationForOrientationMask:orientationMask];
+    if (newOrientation != UIInterfaceOrientationUnknown) {
+      UM_WEAKIFY(self);
+      dispatch_async(dispatch_get_main_queue(), ^{
+        UM_STRONGIFY(self);
+        self.currentScreenOrientation = newOrientation;
+        [[UIDevice currentDevice] setValue:@(newOrientation) forKey:@"orientation"];
```

Comment:
According to [this StackOverflow answer](https://stackoverflow.com/a/20987403/1123156),
> You need to call attemptRotationToDeviceOrientation (UIViewController) to make the system call your supportedInterfaceOrientations when the condition has changed.

so if we set appropriate mask to be returned by the registry and then call `-(void)attemptRotationToDeviceOrientation`, we shouldn't need to set the device's orientation like that. Would it work without it?

([Comment](https://stackoverflow.com/questions/20987249/how-do-i-programmatically-set-device-orientation-in-ios-7#comment31574131_20987403) to this answer suggests this doesn't work, but it's worth checking.)